installinator/src/write.rs

```diff
@@ -0,0 +1,173 @@
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::time::Duration;
+
+use anyhow::{Context, Result};
+use buf_list::BufList;
+use bytes::Buf;
+use camino::{Utf8Path, Utf8PathBuf};
+use installinator_common::{CompletionEventKind, ProgressEventKind};
+use omicron_common::update::ArtifactHashId;
+use tokio::{io::AsyncWriteExt, sync::mpsc, time::Instant};
+
+use crate::reporter::ReportEvent;
+
+#[derive(Clone, Debug)]
+pub(crate) struct WriteDestination {
+    pub(crate) host_phase_2: Utf8PathBuf,
+    pub(crate) control_plane: Utf8PathBuf,
+}
+
+impl WriteDestination {
+    pub(crate) fn in_directory(dir: &Utf8Path) -> Result<Self> {
+        std::fs::create_dir_all(&dir)
+            .with_context(|| format!("error creating directories at {dir}"))?;
+        Ok(Self {
+            host_phase_2: dir.join("host_phase_2.bin"),
+            control_plane: dir.join("control_plane.bin"),
+        })
+    }
+}
+
+pub(crate) async fn write_artifact(
+    log: &slog::Logger,
+    artifact_id: &ArtifactHashId,
+    artifact: BufList,
+    destination: &Utf8Path,
+    event_sender: &mpsc::Sender<ReportEvent>,
+) {
+    let mut attempt = 0;
+
+    loop {
+        attempt += 1;
+        slog::info!(
+            log,
+            "writing artifact ({} bytes) to {destination} (attempt {attempt})",
+            artifact.num_bytes();
+            "artifact_id" => ?artifact_id
+        );
+        match write_artifact_impl(
+            attempt,
+            artifact_id,
+            artifact.clone(),
+            destination,
+            event_sender,
+        )
+        .await
+        {
+            Ok(()) => break,
+            Err(error) => {
+                slog::info!(log, "{error:?}"; "artifact_id" => ?artifact_id);
+                // Give it a short break, then keep trying.
+                tokio::time::sleep(Duration::from_secs(5)).await;
+            }
+        }
+    }
+}
+
+async fn write_artifact_impl(
+    attempt: usize,
+    artifact_id: &ArtifactHashId,
+    artifact: BufList,
+    destination: &Utf8Path,
+    event_sender: &mpsc::Sender<ReportEvent>,
+) -> Result<()> {
+    let mut file = tokio::fs::OpenOptions::new()
+        // TODO: do we want create = true? Maybe only if writing to a file and not an M.2.
+        .create(true)
+        .write(true)
+        .truncate(true)
+        .open(destination)
+        .await
+        .with_context(|| {
+            format!("failed to open destination `{destination}` for writing")
+        })?;
+
+    let total_bytes = artifact.num_bytes() as u64;
+    let mut written_bytes = 0u64;
+
+    let start = Instant::now();
+
+    while artifact.has_remaining() {
+        // BufWriter shouldn't be necessary because we've downloaded, and are
+        // writing, data in typical buffer-sized chunks. Besides,
+        // tokio::fs::File already uses a buffer internally as of tokio 1.25.0.
+        match file.write(artifact.chunk()).await {
+            Ok(n) => {
```

Comment:
I believe it's possible that `file.write` doesn't write the entire contents of `artifact.chunk()`. We should ensure that the entire chunk is written. 

From https://docs.rs/tokio/latest/tokio/io/trait.AsyncWriteExt.html#method.write

> This function will attempt to write the entire contents of buf, but the entire write may not succeed, or the write may also generate an error.

> It is not considered an error if the entire buffer could not be written to this writer.

We could use `write_all` but that is not cancellation safe. It's unclear if that matters here.

Reply:
Oh I need to call advance(n), totally forgot haha

Reply:
I should probably also write a test for this

Reply:
Switched this to `file.write_buf` which already does the advance, and wrote some tests for reliability.